app/bundles/FormBundle/Controller/PublicController.php

```diff
@@ -402,18 +402,28 @@ public function embedAction(Request $request)
      */
     private function replacePostSubmitTokens($string, SubmissionEvent $submissionEvent): string|array
     {
-        if (empty($this->tokens)) {
-            if ($lead = $submissionEvent->getLead()) {
-                $this->tokens = array_merge(
-                    $submissionEvent->getTokens(),
-                    TokenHelper::findLeadTokens(
-                        $string,
-                        $lead->getProfileFields()
-                    )
-                );
-            }
+        if (count($this->tokens)) {
+            return $this->tokens;
+        }
+
+        if ($lead = $submissionEvent->getLead()) {
+            $this->tokens = array_merge(
+                $submissionEvent->getTokens(),
+                TokenHelper::findLeadTokens(
+                    $string,
+                    $lead->getProfileFields()
+                )
+            );
         }
 
+        /** @var \Mautic\PageBundle\Helper\TokenHelper */
+        $pageTokenHelper = $this->factory->get('mautic.page.helper.token');
```

Comment:
Since Mautic 5 you can use controller method autowiring. Just add the service as a typed param and Symfony will prepare it for you:

https://symfony.com/doc/5.x/controller.html#fetching-services

Reply:
This has to be done in `previewAction` which would change its method signature. That'd be a BC. Thus, I avoided that so that this feature can be implemented in M5 and doesn't wait until M6.

Reply:
We've stopped taking that as a BC break in M5 as it became similar to `__construct()` in other services. We should have make the controllers `final` in M5 to make it official but that will have to come in M6. But I appreciate the thoughtfulness!